libvmaf/src/framesync.c

```diff
@@ -0,0 +1,227 @@
+ *     You may obtain a copy of the License at
+ *
+ *         https://opensource.org/licenses/BSDplusPatent
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <pthread.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include "framesync.h"
+
+enum {
+    BUF_FREE = 0,
+    BUF_ACQUIRED,
+    BUF_FILLED,
+    BUF_RETRIEVED,
+};
+
+typedef struct VmafFrameSyncBuf {
+    void *frame_data;
+    int buf_status;
+    signed long index;
+    struct VmafFrameSyncBuf *next;
+} VmafFrameSyncBuf;
+
+typedef struct VmafFrameSyncContext {
+    VmafFrameSyncBuf *buf_que;
+    pthread_mutex_t acquire_lock;
+    pthread_mutex_t retrieve_lock;
+    pthread_cond_t retrieve;
+    unsigned buf_cnt;
+} VmafFrameSyncContext;
+
+int vmaf_framesync_init(VmafFrameSyncContext **fs_ctx)
+{
+    VmafFrameSyncContext *const ctx = *fs_ctx = malloc(sizeof(VmafFrameSyncContext));
+    if (!ctx) return -ENOMEM;
+    memset(ctx, 0, sizeof(VmafFrameSyncContext));
+    ctx->buf_cnt = 1;
+
+    pthread_mutex_init(&(ctx->acquire_lock), NULL);
+    pthread_mutex_init(&(ctx->retrieve_lock), NULL);
+    pthread_cond_init(&(ctx->retrieve), NULL);
+
+    VmafFrameSyncBuf *buf_que = ctx->buf_que = malloc(sizeof(VmafFrameSyncBuf));
+
+    buf_que->frame_data = NULL;
+    buf_que->buf_status = BUF_FREE;
+    buf_que->index = -1;
+    buf_que->next  = NULL;
+
+    return 0;
+}
+
+int vmaf_framesync_acquire_new_buf(VmafFrameSyncContext *fs_ctx, void **data,
+                                   unsigned data_sz, unsigned index)
+{
+    VmafFrameSyncBuf *buf_que = fs_ctx->buf_que;
+    *data = NULL;
+
+    pthread_mutex_lock(&(fs_ctx->acquire_lock));
+
+    // traverse until a free buffer is found
+    for (unsigned i = 0; i < fs_ctx->buf_cnt; i++) {
+        if (buf_que->buf_status == BUF_FREE) {
+            buf_que->frame_data = *data = malloc(data_sz);
+            if (!buf_que->frame_data)
+                return -ENOMEM;
+            buf_que->buf_status = BUF_ACQUIRED;
+            buf_que->index = index;
+            break;
+        }
+        // move to next node
+        if (buf_que->next != NULL)
+            buf_que = buf_que->next;
+    }
+
+    // create a new node if all nodes are occupied in the list and append to the tail
+    if (*data == NULL) {
+        VmafFrameSyncBuf *new_buf_node = malloc(sizeof(VmafFrameSyncBuf));
+        buf_que->next = new_buf_node;
+        new_buf_node->buf_status = BUF_FREE;
+        new_buf_node->index = -1;
+        new_buf_node->next = NULL;
+        fs_ctx->buf_cnt++;
+
+        new_buf_node->frame_data = *data = malloc(data_sz);
```

Comment:
nit: There is some duplicated logic here, maybe it is possible to extract out the "create new node" and "allocate node's data" into two functions to be reused. But we can refactor this later.

Reply:
Yes, agreed with 2 functions the code can made reusable.